### DISCUSSION TITLE
Is there currently a way to parse a raw ClassName from a textual representation, or a KtTypeReference?

### DISCUSSION THREAD

**Discussion by Sintrastes** [square/kotlinpoet#1187] [Q&A] [answered]
I am building a compiler plugin using kotlinpoet -- and in my generated code, some of the types of the source code need to pass through unchanged. I have access to these types in the form of a `org.jetbrains.kotlin.psi.KtTypeReference` -- but I cannot for the life of me figure out how to parse such a type into a kotlinpoet `ClassName`.

If I just use the raw text of the `KtTypeReference`, composite types like `List<Int>` give me issues, and nullable types come out looking like:

    `Double?`

rather than simply

    Double?

Is there currently a way to translate/parse a `KtTypeReference` into a `ClassName` in kotlinpoet, either directly from the Kotlin PSI API, or via parsing raw text? I could write a parser myself, but I'd like to avoid re-inventing the wheel.

**Comment by ZacSweers** [comment, marked as answer]
We don't have a first party PSI artifact, but https://github.com/square/anvil has a bunch of utilities around this and almost certainly has some prior art you could use.

**Reply by Sintrastes** [reply]
Thanks! Looks like `KotlinPoetUtils.kt` has what I need. 

Would be nice if there were a specific artifact for this (`kotlinpoet-psi-utils`?), but for now I should be able to adapt the source file to my project, under the appropriate license constraints of course.
